File: built-in plugins dsp/mb-drums-acoustic.cpp
/**
 * MB Acoustic Kit
 * Category : instrument
 * Type     : drums
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Natural acoustic drum kit
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_DRUMS_ACOUSTIC_H
#define MB_DRUMS_ACOUSTIC_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbDrumsAcoustic : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-drums-acoustic";
    static constexpr const char* PLUGIN_NAME    = "MB Acoustic Kit";
    static constexpr const char* PLUGIN_TYPE    = "drums";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float punch = 0.7f;  // range [0, 1]
    float room = 0.3f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbDrumsAcoustic() = default;
    ~MbDrumsAcoustic() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.punch = std::clamp(params.punch, 0f, 1f);
        params.room = std::clamp(params.room, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Acoustic Kit
        return input;
    }
};

#endif // MB_DRUMS_ACOUSTIC_H

File: built-in plugins dsp/mb-drums-acoustic.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbDrumsAcousticPlugin: PluginDefinition = {
    id: 'mb-drums-acoustic', slug: 'mb-drums-acoustic', name: 'MB Acoustic Kit', category: 'instrument', type: 'drums', version: '1.0.0',
    description: 'Natural acoustic drum kit', author: 'Max Booster', grade: 'A',
    oscillators: [{ type: 'sine', detune: 0, gain: 1.0 }, { type: 'noise', detune: 0, gain: 0.5 }],
    envelope: { attack: 0.001, decay: 0.2, sustain: 0.0, release: 0.3 },
    parameters: [
      { id: 'punch', name: 'Punch', type: 'float', defaultValue: 0.7, minValue: 0, maxValue: 1, automatable: true },
      { id: 'room', name: 'Room', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { punch: 0.7, room: 0.3, volume: 0.8 },
  };

export default MbDrumsAcousticPlugin;
